apps/wallet/src/background/account-sources/QredoAccountSource.ts

```diff
@@ -0,0 +1,180 @@
+	}: {
+		password: string;
+		apiUrl: string;
+		organization: string;
+		origin: string;
+		service: string;
+		refreshToken: string;
+	}) {
+		const decryptedData: DataDecryptedV0 = {
+			version: 0,
+			refreshToken,
+		};
+		const dataSerialized: QredoAccountSourceSerialized = {
+			id: makeUniqueKey(),
+			type: 'qredo',
+			apiUrl,
+			organization,
+			origin,
+			service,
+			encrypted: await encrypt(password, decryptedData),
+		};
+		const allAccountSources = await getAccountSources();
+		for (const anAccountSource of allAccountSources) {
+			if (
+				anAccountSource instanceof QredoAccountSource &&
+				isSameQredoConnection(
+					{
+						apiUrl: await anAccountSource.apiUrl,
+						organization: await anAccountSource.organization,
+						origin: await anAccountSource.origin,
+						service: await anAccountSource.service,
+					},
+					{ id: dataSerialized.id, apiUrl, organization, origin, service },
+				)
+			) {
+				throw new Error('Qredo account source already exists');
+			}
+		}
+		await (await getDB()).accountSources.put(dataSerialized);
+		await backupDB();
+		accountSourcesEvents.emit('accountSourcesChanged');
+		return new QredoAccountSource(dataSerialized.id);
+	}
+
+	static isOfType(serialized: AccountSourceSerialized): serialized is QredoAccountSourceSerialized {
+		return serialized.type === 'qredo';
+	}
+
+	async toUISerialized(): Promise<QredoAccountSourceSerializedUI> {
+		const { apiUrl, id, organization, origin, service } = await this.getStoredData();
+		return {
+			id,
+			type: 'qredo',
+			origin,
+			apiUrl,
+			organization,
+			service,
+			isLocked: await this.isLocked(),
+		};
+	}
+
+	async isLocked(): Promise<boolean> {
+		return !(await this.getEphemeralValue());
+	}
+
+	async lock(): Promise<void> {
+		await this.clearEphemeralValue();
+		accountSourcesEvents.emit('accountSourceStatusUpdated', { accountSourceID: this.id });
+	}
+
+	async unlock(password: string) {
+		const { encrypted } = await this.getStoredData();
+		const { refreshToken } = await decrypt<DataDecryptedV0>(password, encrypted);
+		await this.setEphemeralValue({
+			refreshToken,
+			accessToken: await this.#createAccessToken(refreshToken),
+		});
+		accountSourcesEvents.emit('accountSourceStatusUpdated', { accountSourceID: this.id });
+	}
+
+	async renewAccessToken() {
+		const ephemeralData = await this.getEphemeralValue();
+		if (!ephemeralData) {
+			throw new Error(`Qredo account source ${this.id} is locked`);
+		}
+		const { refreshToken } = ephemeralData;
+		const accessToken = await this.#createAccessToken(refreshToken);
+		await this.setEphemeralValue({ refreshToken, accessToken });
+		return accessToken;
+	}
+
+	async #createAccessToken(refreshToken: string): Promise<string> {
+		const { apiUrl } = await this.getStoredData();
+		return (
+			await new QredoAPI(this.id, apiUrl).createAccessToken({
+				refreshToken,
+			})
+		).access_token;
+	}
+
+	get apiUrl() {
+		return this.getStoredData().then(({ apiUrl }) => apiUrl);
```

Comment:
we don't use `cachedData` here is that intended?

Reply:
Yeah cache data for accounts was done mainly as optimisation and here it didn't seem necessary. (we can do it later if necessary)